presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java

```diff
@@ -184,6 +187,18 @@ public ListenableFuture<?> isBlocked()
     @Override
     public boolean needsInput()
     {
+        // We can skip probe for empty build input only when probeOnOuterSide is false
+        if (optimizeProbeForEmptyBuild && !probeOnOuterSide) {
+            if (tryFetchLookupSourceProvider()) {
+                lookupSourceProvider.withLease(lookupSourceLease -> {
+                    // Do not have spill, build side is empty and probe side does not output for non match, skip and finish the operator
+                    if (!lookupSourceLease.hasSpilled() && lookupSourceLease.getLookupSource().isEmpty()) {
+                        finish();
+                    }
+                    return null;
+                });
+            }
+        }
```

Comment:
This is the core logic, other code is to pass the optimizeProbeForEmptyBuild flag.

Reply:
This looks bit messy to me. Why spilling matters in this case if it is already empty?

Reply:
When we reach to this code, i.e. `tryFetchLookupSourceProvider` returns true, the build of hash table is already complete, i.e. spill has already complete if any.

For spilled partitions, the exact number of rows is not available until unspilling, and the assumption is that it's always not empty. I added the check here for additional safety, i.e. with spilling it will not be empty and we cannot skip probe here.